transformer_lens/utils.py

```diff
@@ -807,3 +807,141 @@ def override_or_use_default_flag(
     Otherwise, the global flag is returned.
     """
     return override if override is not None else default_flag
+
+
+def get_attention_mask(
+    tokenizer,
+    tokens: Int[torch.Tensor, "batch pos"],
+    prepend_bos: bool,
+) -> Int[torch.Tensor, "batch pos"]:
+    """
+    This method calculates an attention mask for a given sequence of tokens when left padding is used. The attention
+    mask is a binary tensor where each element is 1 if the corresponding token is not a pad token and 0 otherwise. In
+    the special case where the BOS token is the same as the pad token and bos token is prepended to process the tokens,
+    this method ensures that the prepended BOS (pad) token also gets a 1 in the attention mask.
+    Gotcha: Make sure that the passed prepend_bos is the same as the one used to generate the tokens!
+    """
+    # Initialize the attention mask to be 1 wherever the token is not a pad token
+    attetnion_mask = tokens.ne(tokenizer.pad_token_id)
+
+    # Handle the special case where the BOS token is the same as the pad token
+    if tokenizer.bos_token_id == tokenizer.pad_token_id and prepend_bos:
+        if tokenizer.padding_side == "right":
+            # Set the first token to 1 in the attention mask
+            attetnion_mask[:, 0] = True
+
+        else:
+            is_pad_token = 1 - attetnion_mask.int()
+
+            # Find the position of the pad token used as the BOS token and thus should get attended
+            pad_bos_positions = is_pad_token.cumsum(dim=-1).argmax(dim=-1)
```

Comment:
I believe this will break when the BOS EOS *and* PAD tokens are the same, as is the case in eg GPT-2. It would be good to add a test for this case. EOS is often used to separate documents in the same context, eg "<PAD><PAD><BOS>The cat sat on the mat<EOS>The dog sat on the log". I would add a separate check that the pad tokens are contiguous (eg I believe ((is_pad_token.cumsum(-1) * ((1-is_pad_token).cumsum(-1)>0)).argmax(dim=-1) should work?) I think you should also check for the case where there's no PAD/BOS/EOS in the context, and add a test for correct handling. And this can be different by row, ugh.

This is very fiddly, sorry for all the nitpicks!

Reply:
Right, the assumption here was that all the PAD tokens are contiguously at the beginning and no PAD token appears in the middle. I didn't consider the case of PAD token being an EOS token and appearing at the middle of the text. The suggestion of checking if they are contiguous is great. Thank you so much for pointing out the exceptional case!

Reply:
Resolved by commit 0591924a63b2591a02b84cfa6f138d0c4303a1b5

Reply:
Note: 0591924a63b2591a02b84cfa6f138d0c4303a1b5 always considers (regardless of whether pad == bos or pad == sep) _only_ the leftmost leading pads or rightmost trailing pads as _real_ pad tokens that should not be attended.
Rational: Sep tokens should be attended, and there might be the cases where users use sep tokens multiple times.